src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,414 @@
+  { rcases z with ⟨z, hz2⟩,
+    rw subtype.image_preimage_coe at hz2,
+    exact hz2.2, },
+  { exact ideal.subset_span z.2 },
+end
+
+variables {𝒜 I}
+
+lemma ideal.is_homogeneous.homogeneous_core_eq_self (h : I.is_homogeneous 𝒜) :
+  I = I.homogeneous_core 𝒜 :=
+begin
+  apply le_antisymm _ (I.homogeneous_core_le_ideal 𝒜),
+  intros x hx,
+  letI : Π (i : ι) (x : 𝒜 i), decidable (x ≠ 0) := λ _ _, classical.dec _,
+  rw ←graded_algebra.sum_support_decompose 𝒜 x,
+  refine ideal.sum_mem _ _,
+  intros j hj,
+  apply ideal.subset_span,
+  exact ⟨⟨_, is_homogeneous_coe _⟩, h _ hx, rfl⟩,
+end
+
+variables (𝒜 I)
+
+lemma ideal.is_homogeneous.iff_eq :
+  I.is_homogeneous 𝒜 ↔ I = I.homogeneous_core 𝒜 :=
+⟨ λ hI, hI.homogeneous_core_eq_self,
+  λ hI, hI.symm ▸ I.is_homogeneous_homogeneous_core 𝒜 ⟩
+
+lemma ideal.is_homogeneous.iff_exists :
+  I.is_homogeneous 𝒜 ↔ ∃ (S : set (homogeneous_submonoid 𝒜)), I = ideal.span (coe '' S) :=
+by rw [ideal.is_homogeneous.exists_iff_eq_span, ideal.is_homogeneous.iff_eq]
+
+end is_homogeneous_ideal_defs
+
+section operations
+
+variables {ι R A : Type*} [comm_ring R] [comm_ring A] [algebra R A]
+variables [decidable_eq ι] [add_comm_monoid ι]
+variables (𝒜 : ι → submodule R A) [graded_algebra 𝒜]
+variable (I : ideal A)
+
+/--For any `comm_ring R`, we collect the homogeneous ideals of `R` into a type.-/
+abbreviation homogeneous_ideal : Type* := { I : ideal A // I.is_homogeneous 𝒜 }
+
+lemma ideal.is_homogeneous.bot : ideal.is_homogeneous 𝒜 ⊥ := λ i r hr,
+begin
+  simp only [ideal.mem_bot] at hr,
+  rw [hr, alg_equiv.map_zero, zero_apply],
+  apply ideal.zero_mem
+end
+
+instance homogeneous_ideal.inhabited : inhabited (homogeneous_ideal 𝒜) :=
+{ default := ⟨⊥, ideal.is_homogeneous.bot _⟩}
+
+instance homogeneous_ideal.has_top :
+  has_top (homogeneous_ideal 𝒜) :=
+⟨⟨⊤, λ _ _ _, by simp only [submodule.mem_top]⟩⟩
+
+@[simp] lemma homogeneous_ideal.eq_top_iff (I : homogeneous_ideal 𝒜) : I = ⊤ ↔ I.1 = ⊤ :=
+subtype.ext_iff
+
+instance homogeneous_ideal.order : partial_order (homogeneous_ideal 𝒜) :=
+partial_order.lift _ subtype.coe_injective
+
+instance homogeneous_ideal.has_mem : has_mem A (homogeneous_ideal 𝒜) :=
+{ mem := λ r I, r ∈ I.1 }
+
+variables {𝒜}
+
+lemma ideal.is_homogeneous.inf {I J : ideal A}
+  (HI : I.is_homogeneous 𝒜) (HJ : J.is_homogeneous 𝒜) : (I ⊓ J).is_homogeneous 𝒜 :=
+λ i r hr, ⟨HI _ hr.1, HJ _ hr.2⟩
+
+lemma ideal.is_homogeneous.Inf {ℐ : set (ideal A)} (h : ∀ I ∈ ℐ, ideal.is_homogeneous 𝒜 I) :
+  (Inf ℐ).is_homogeneous 𝒜 :=
+begin
+  intros i x Hx, simp only [ideal.mem_Inf] at Hx ⊢,
+  intros J HJ,
+  exact h _ HJ _ (Hx HJ),
+end
+
+lemma ideal.is_homogeneous.mul {I J : ideal A}
+  (HI : I.is_homogeneous 𝒜) (HJ : J.is_homogeneous 𝒜) : (I * J).is_homogeneous 𝒜 :=
+begin
+  rw ideal.is_homogeneous.iff_exists at HI HJ ⊢,
+  obtain ⟨⟨s₁, rfl⟩, ⟨s₂, rfl⟩⟩ := ⟨HI, HJ⟩,
+  rw ideal.span_mul_span',
+  refine ⟨s₁ * s₂, congr_arg _ _⟩,
+  exact (set.image_mul (submonoid.subtype _).to_mul_hom).symm,
+end
```

Comment:
I was able to golf this substantially by looking for an existing lemma rather than trying to unfold everything